### DISCUSSION TITLE
Dynamic import() (lazy loading) to reduce bundle size

### DISCUSSION THREAD

**Discussion by christiankaindl** [algolia/instantsearch#5333] [Q&A]
Hi, I am using `react-instantsearch-hooks-web` in my Next.js project and have a question/feature request. We use algolia for our search dropdown and on some pages for search results. I want to use Next.js' dynamic imports to reduce my bundle size and improve loading performance, as the main imports are quite heavy:
<img width="856" alt="Screen Shot 2022-09-21 at 12 39 01" src="https://user-images.githubusercontent.com/15364860/191483639-7f5e3031-ed2e-4d91-b236-13bfbb3f44d9.png">

Which is about 30% of my total bundle size. Considering that the search experience is not important on first load, but only after users interact with the search, I want to lazy load it. I tried using Next.js' `dynamic()` function and skip rendering the `InstantSearch` component while loading, but then I get errors because the hooks expect the context providers to be present:

<img width="986" alt="Screen Shot 2022-09-21 at 12 41 10" src="https://user-images.githubusercontent.com/15364860/191484065-9c81d1dc-81d9-41fc-b336-9d8f1dcde301.png">

This is roughly the code I tried:
```jsx
import algoliasearch from 'algoliasearch/lite'
import { FC, Suspense } from 'react'
import dynamic from 'next/dynamic'

// Lazily import Algolia components
const hooksWeb = import('react-instantsearch-hooks-web')
const InstantSearch = dynamic(async () => await hooksWeb.then((mod) => mod.InstantSearch), {
  ssr: true,
  suspense: true
})
const Configure = dynamic(async () => await hooksWeb.then((mod) => mod.Configure), {
  ssr: true,
  suspense: true
})

// Set up algolia search client (would be nice to lazy load as well...)
export const searchClient = algoliasearch(process.env.NEXT_PUBLIC_ALGOLIA_APP_ID as string, process.env.NEXT_PUBLIC_ALGOLIA_API_KEY as string)

const SearchProvider: FC<React.PropsWithChildren> = function ({ children }) {
  return (
      // Use Suspense to only render the children while InstantSearch is loading
      <Suspense fallback={children}>
        <InstantSearch searchClient={searchClient} indexName={process.env.NEXT_PUBLIC_ALGOLIA_SUGGESTIONS_INDEX}>
          <Configure
            // config props...
          />
          {children}
        </InstantSearch>
      </Suspense>
  )
}
export default SearchProvider
```

One approach could be to wrap the fallback in a `InstantSearchContext` and leave its value empty, but `InstantSearchContext` is currently not exported...

**Comment by christiankaindl** [comment]
If I could import `InstantSearchContext` like
```js
import { InstantSearchContext } from 'react-instantsearch-hooks/dist/es/lib/InstantSearchContext'
```
and use it as fallback (with mocked `value`), that would already be enaugh for our use-case. That's currently not possible because it is not exported in package.json

**Comment by Haroenv** [comment]
Could the answer not be to lazy load the whole search component like this:

```js
const Search = dynamic(
  async () => {
    const { InstantSearch, Configure } = await import(
      'react-instantsearch-hooks-web'
    );
    return () => (
      <InstantSearch indexName="indexName" searchClient={searchClient}>
        <Configure hitsPerPage={3} />
        {/*...*/}
      </InstantSearch>
    );
  },
  {
    ssr: true,
    suspense: true,
  }
);

const SearchProvider: FC<React.PropsWithChildren> = function ({ children }) {
  return (
    // Use Suspense to only render the children while InstantSearch is loading
    <Suspense fallback={children}>
      <Search />
    </Suspense>
  );
};
```

If you will mock the InstantSearch provider, you're basically going to rewrite all of InstantSearch or create state mismatch bugs easily, which is why the context isn't exposed.

**Reply by christiankaindl** [reply]
Thanks for your reply! No, that doesn't work sadly. If any of the components in `children` is using a hook of `react-instantsearch-hooks-web` the application breaks because they need the context to be available (see screenshot in the OP). That's the whole problem :) So in your example, the `children` in the Suspense fallback would need some sort of Context Provider to not break the `react-instantsearch-hooks` hooks

Ideally there would be some sort of wrapper component available that causes hooks to either return empty, or return any server generated data (collected with `getServerState()` and `InstantSearchSSRProvider`), as long as the actual InstantSearch component is still loading.

**Reply by Haroenv** [reply]
So if I understand it correctly, you only want to lazy load the provider of the data, not anything else? I assumed you'd want to lazy load everything else as well

**Reply by christiankaindl** [reply]
I want to lazyload the `InstantSearch` and `Configure` components as they are quite large in terms of bundle size. What do you mean by "just the provider" of the data?

Your code example and mine are roughly equal, no? Load the components via `import()` and use Suspense to show a fallback while loading. The issue is that the fallback being `children` doesn't work if any of the children use a `react-instantsearch-hooks` hook (like `useHits()` or `useStats()`). I it helps I will come back with a Sandbox reproduction of the issue :)

**Reply by christiankaindl** [reply]
Just to clarify: the example you provided does work, but only if none of the child components are using any InstantSearch hooks like useHits(). That is because useHits() needs the context provided by InstantSearch

**Reply by christiankaindl** [reply]
> If you will mock the InstantSearch provider, you're basically going to rewrite all of InstantSearch or create state mismatch bugs easily, which is why the context isn't exposed.

What about when using InstantSearchSSRProvider, like from your docs here https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering/react-hooks/#server-side-rendering? The data is not even mocked in that case, it is an actual InstantSearch state object :)

**Comment by jdriscoll98** [comment]
@christiankaindl Hey Christian! This is quite an old issue but I'm in the same boat as you right now. Did you ever figure out how to do this?

**Reply by christiankaindl** [reply]
No, not the way I had hoped for. But there's two things we did to improve our implementation:
- We since added server side rendering (see [this guide](https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering/react-hooks/)), so that the initial page load already displays useful content
- We moved our `<InstantSearch>` usage from the _app.tsx file to each page that needs it, so that pages without search don't incur the bundle size cost. Additionally, we now lazy load our search dropdown (via NextJS' dynamic() function) to reduce initial load time.

**Edit**: Alternatively you could remove InstantSearch from the client completely and only use it in the backend, and expose that through an API endpoint
